### PR TITLE
Add save to file functionality.

### PR DESCRIPTION
**Pull request by duckonomy**
## Overview: ##
Adds exporting to tapis systems and linking with DS projects. 

## PR Status: ##

* [X] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

* [DES-1929](https://jira.tacc.utexas.edu/browse/DES-1929)

## Summary of Changes: ##
Added three different scenarios of "saving/exporting", The first is exporting a file for an existing map. If you select a project system, it will allow you to link to that project. The second case is when creating a project. You will be able to link to a project system and create the file (in its root), The third case is for observable projects (assuming that the observable project is in a DS project). When creating, it should automatically create the link and the file.

## Testing Steps: ##
0. Use the geoapi branch that supports this (TACC-Cloud/geoapi#54)
### Exporting
1. After creating a project (without the link option)
2. Click the "Export map" button in the users/manage panel. 
3. Select a system & folder,
4. Confirm that a `<uuid>.hazmapper` file of the current project is created. 
5. Try again selecting a project system and confirm that the file and a link to that project is created (on the top left control-bar and on the manage panel.
6. Also try creating with an already exported map.
7. This should prompt you to confirm that the file gets removed in the previously saved location also the link to the project.

### Creating a map
1. Try creating a project with the "Export Project" checked
2. After creation, confirm that the map is linked to the selected project.

### Creating an observable project
1. Create an observable map from a rapid project in a DS project.
2. Confirm that both file and link are created. 

## UI Photos:
Screenshot_2021-04-21 HazMapper
localhost_4200_ (2)
localhost_4200_ (1)
localhost_4200_

### CODE REVIEW

File: src/app/components/main-welcome/main-welcome.component.styl [hunk 1]
```diff
@@ -14,15 +14,16 @@
     #user-guide
       width: 550px
       text-align: right
-    
+  .list-column
```
Comment (nathanfranklin) [inline src/app/components/main-welcome/main-welcome.component.styl:17]: could we make the `welcome-projects-list` wider (width:100%) and then maybe make these columns take up that space (but give less space to the 3rd column).  maybe a follow on ticket?

File: src/app/components/modal-create-project/modal-create-project.component.html [hunk 2]
```diff
@@ -22,8 +64,10 @@ <h4> Create a new Map </h4>
           [onlyFolder]="true"
           [single]="true"
           [heading]="'Select a RApp Folder'"
+          [allowedExtensions]="allowedExtensions"
```
Comment (nathanfranklin) [inline src/app/components/modal-create-project/modal-create-project.component.html:67]: is this change needed?

File: src/app/components/modal-create-project/modal-create-project.component.ts [hunk 3]
```diff
@@ -14,33 +14,55 @@ import {Subject} from 'rxjs';
...
 export class ModalCreateProjectComponent implements OnInit {
 
-  public readonly onClose: Subject<any> = new Subject<any>();
-
   projCreateForm: FormGroup;
   rapidFolder: RemoteFile;
   submitting: boolean;
   errorMessage = '';
+  currentPath: string;
+  fileSuffix: string = '';
```
Comment (nathanfranklin) [inline src/app/components/modal-create-project/modal-create-project.component.ts:22]: should this be renamed to "fileName" or "fileBaseName" as the suffix is always "hazmapper".

**Review by nathanfranklin — CHANGES_REQUESTED** [review]
Looks great!  Left some comments to consider.

Lots of my comments have suggestions for renaming things.  I'm not certain though if my suggestions are any good.  So I don't know if you (and @reptillicus ) consider what I wrote and if names need to be changed.  Or if we can poll Craig for his opinion.  Or just wait till we get feedback from Tracy.

File: src/app/components/modal-link-project/modal-link-project.component.html [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  <label> Custom file name:
+    <input type="text" [(ngModel)]="fileSuffix">
+  </label>
+  <small><em>If name is not given, file will be saved as the map uuid.</em></small>
+</div>
+
+<div class="confirm-checkbox">
+  <label>
+    <span [ngClass]="{dimmed: linkCheckbox.disabled}">Link Project</span>
+    <input #linkCheckbox type="checkbox"
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/components/modal-link-project/modal-link-project.component.html:25]: I've checked it by default if the user is in a project folder.

File: src/app/components/modal-create-project/modal-create-project.component.html [hunk 5]
```diff
@@ -10,9 +10,51 @@ <h4> Create a new Map </h4>
...
+                (currentPath)="setCurrentPath($event)"
+                (systemSelection)="onSystemSelect($event)">
+              </app-file-browser>
+
+              <div style="padding: 20px 0px">
+                <label> Custom file name:
+                  <input type="text"
+                         formControlName="fileSuffix">
+                </label>
+                <small><em>If name is not given, file will be saved as the map uuid.</em></small>
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/components/modal-create-project/modal-create-project.component.html:41]: 1. For the link modal, I can do that. But for the create project, I wouldn't have the UUID beforehand.
2. I will try doing this and create a separate Ticket/PR for that feature.

File: src/app/components/users-panel/users-panel.component.html [hunk 6]
```diff
@@ -34,6 +34,27 @@ <h6 id="property-title">Description:</h6>
...
+      </div>
+      <ng-template #noSave>
+        <div style="margin-bottom: 20px">
+          Map not saved.
+        </div>
+      </ng-template>
+
+      <h6 id="property-title">Linked Project</h6>
+      <div *ngIf="activeProject.system_name && activeProject.system_id.includes('project'); else noLink" style="margin-bottom: 20px">
+        <a [href]="'https://www.designsafe-ci.org/data/browser/projects/' + activeProject.system_id + '/'"><span>{{activeProject.system_name}}</span></a>
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/components/users-panel/users-panel.component.html:49]: I'm not sure where to get the PRJ-XXXX (as displayed in DS). The ID in tapis seemed to be a much longer one.

[the remark this answers is not in the document]
Comment (nathanfranklin) [reply src/app/components/users-panel/users-panel.component.html:49]: It is in the projects response of https://agave.designsafe-ci.org//projects/v2/.  BUT yeah, that is probably not worth the effort at this stage (when we don't have Tracy feedback).  So maybe just ignore this comment.

File: src/app/components/main-welcome/main-welcome.component.html [hunk 7]
```diff
@@ -40,22 +43,33 @@
...
-                  <div>
+
+                  <div class="list-column">
+                    {{p.name}}
+                  </div>
+
+                  <div class="list-column">
+                    <ng-container *ngIf="p.system_name && p.system_id.includes('project')" class="list-column link-hover">
+                      <a [href]="'https://www.designsafe-ci.org/data/browser/projects/' + p.system_id + '/'">
+                        <span>{{p.system_name}}</span>
```
[the remark this answers is not in the document]
Comment (nathanfranklin) [reply src/app/components/main-welcome/main-welcome.component.html:63]: Ignore due to https://github.com/TACC-Cloud/hazmapper/pull/50#discussion_r632779256

File: src/app/components/users-panel/users-panel.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/components/users-panel/users-panel.component.html]: Gotcha!

File: src/app/components/modal-create-project/modal-create-project.component.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (nathanfranklin) [reply src/app/components/modal-create-project/modal-create-project.component.html:41]: gotcha 👍

File: src/app/components/users-panel/users-panel.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/components/users-panel/users-panel.component.html]: Got this working!

File: src/app/components/users-panel/users-panel.component.ts [hunk 8]
```diff
@@ -45,6 +67,31 @@ export class UsersPanelComponent implements OnInit {
...
+                                               next.system.id,
+                                               path,
+                                               next.fileSuffix)
+      } else {
+        this.projectsService.exportProject(this.activeProject.id,
+                                           next.system.id,
+                                           path,
+                                           next.fileSuffix)
+      }
+    });
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/components/users-panel/users-panel.component.ts:88]: Should I handle it in the service function?

File: src/app/services/projects.service.ts [hunk 9]
```diff
@@ -96,6 +99,38 @@ export class ProjectsService {
...
+      system_id,
+      path,
+      file_suffix
+    };
+    this.http.put<any>(this.envService.apiUrl + `/projects/${projectId}/export/`, payload).subscribe(current_project => {
+      this.notificationsService.showSuccessToast(`Create file ${system_id}/${path}/${current_project.uuid}.hazmapper`)
+      this._projects.next([...this._projects.value.filter((p) => p.id != projectId),
+                           current_project]);
+    }, error => {
+      console.log(error);
```
[the remark this answers is not in the document]
Comment (duckonomy) [reply src/app/services/projects.service.ts]: I'll add a ticket for error handling. For now I added an error toast.

**Comment by reptillicus** [comment]
Can we pull the "Make map public" button to the left and add a divider or something before the Export Map and Delete buttons ?

File: src/app/components/users-panel/users-panel.component.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (nathanfranklin) [reply src/app/components/users-panel/users-panel.component.ts:99]: I see you handled it there 👍

**Review by nathanfranklin — APPROVED** [review]
Looks good.  💯     The new layout of the Manage component looks very nice; a great improvement!

I think the few comments I have are very minor and can be stuck in a follow on ticket (and in some cases ignored). I guess take a look and see what you think.

**Pull request merged**